File: user-service/src/main/java/rs/raf/user_service/repository/VerificationRequestRepository.java
package rs.raf.user_service.repository;

import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Query;
import org.springframework.data.repository.query.Param;
import rs.raf.user_service.domain.entity.VerificationRequest;

import java.util.List;
import java.util.Optional;

public interface VerificationRequestRepository extends JpaRepository<VerificationRequest, Long> {

    @Query("SELECT v FROM VerificationRequest v WHERE v.userId = :userId AND v.status = 'PENDING' AND v.expirationTime > CURRENT_TIMESTAMP ORDER BY v.createdAt DESC")
    List<VerificationRequest> findActiveRequests(@Param("userId") Long userId);

    @Query("SELECT v FROM VerificationRequest v WHERE v.userId = :userId AND ((v.status != 'PENDING') OR v.expirationTime < CURRENT_TIMESTAMP) ORDER BY v.createdAt DESC")
    List<VerificationRequest> findInactiveRequests(@Param("userId") Long userId);

    @Query("SELECT v FROM VerificationRequest v WHERE v.id = :id AND v.userId = :userId AND v.status = 'PENDING' AND v.expirationTime > CURRENT_TIMESTAMP ORDER BY v.createdAt DESC")
    Optional<VerificationRequest> findActiveRequest(@Param("id") Long id, @Param("userId") Long userId);

}

File: user-service-kotlin/src/main/kotlin/pack/userservicekotlin/repository/VerificationRequestRepository.kt
package pack.userservicekotlin.repository

import org.springframework.data.jpa.repository.JpaRepository
import org.springframework.data.jpa.repository.Query
import org.springframework.data.repository.query.Param
import org.springframework.stereotype.Repository
import pack.userservicekotlin.domain.entities.VerificationRequest

@Repository
interface VerificationRequestRepository : JpaRepository<VerificationRequest, Long> {
    @Query(
        """
        SELECT v FROM VerificationRequest v 
        WHERE v.userId = :userId 
          AND v.status = 'PENDING' 
          AND v.expirationTime > CURRENT_TIMESTAMP 
        ORDER BY v.createdAt DESC
    """,
    )
    fun findActiveRequests(
        @Param("userId") userId: Long,
    ): List<VerificationRequest>

    @Query(
        """
        SELECT v FROM VerificationRequest v 
        WHERE v.userId = :userId 
          AND (v.status != 'PENDING' OR v.expirationTime < CURRENT_TIMESTAMP) 
        ORDER BY v.createdAt DESC
    """,
    )
    fun findInactiveRequests(
        @Param("userId") userId: Long,
    ): List<VerificationRequest>

    @Query(
        """
        SELECT v FROM VerificationRequest v 
        WHERE v.id = :id 
          AND v.userId = :userId 
          AND v.status = 'PENDING' 
          AND v.expirationTime > CURRENT_TIMESTAMP
    """,
    )
    fun findActiveRequest(
        @Param("id") id: Long,
        @Param("userId") userId: Long,
    ): VerificationRequest?
}
